### PR TITLE
fix: windows save dialog extension bug

### LINKED ISSUE #44277: [Bug]: Windows showSaveDialog invalid filename bug
### Preflight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

33.0.0

### What operating system(s) are you using?

Windows

### Operating System Version

Windows 11 22H2 22621.4317

### What arch are you using?

x64

### Last Known Working Electron version

None

### Expected Behavior

1. Open the save dialog using dialog.showSaveDialog.
2. Enter a file name containing illegal characters (e.g., 111*222).
- Expected: The dialog correctly shows a warning that the file name is invalid due to the presence of illegal characters*.
3. Remove the illegal characters* (e.g., change 111*222 to 111222).
4. Attempt to save the file without adding an extension.
- Expected: The dialog should now allow saving the file since all illegal characters have been removed, even if no file extension is added.

### Actual Behavior

1. Open the save dialog using dialog.showSaveDialog.
2. Enter a file name containing illegal characters* (e.g., 111*222).
- Actual: The dialog provides no error message or feedback.
3. Remove the illegal characters* (e.g., change 111*222 to 111222).
4. Attempt to save the file without adding an extension.
- Actual: The dialog shows the file name as invalid, even though all illegal characters have been removed.
5. Now, modify the file name to include a valid extension (e.g., 111222.xml).
- Actual: The file saves successfully when an extension is added, indicating that the dialog behaves inconsistently by requiring an extension even for valid file names.

### Testcase Gist URL

_No response_

### Additional Information

Below is a minimal code example to reproduce the issue:

Sample Code:
```
const fs = require('fs');
const { dialog } = require('electron');

const saveFileDialog = dialog.showSaveDialog({
  title: '保存 XML 文件',
  filters: [
    { name: 'XML 文件', extensions: ['xml'] }
  ]
});

saveFileDialog.then(result => {
  if (!result.canceled) {
    const filePath = result.filePath;
    console.log(`文件将保存到: ${filePath}`);

    // 你要保存的 XML 内容
    const xmlContent = `
<?xml version="1.0" encoding="UTF-8"?>
<root>
  <item>
    <name>示例数据</name>
    <value>123</value>
  </item>
</root>`;

    // 将 XML 内容写入文件
    fs.writeFile(filePath, xmlContent, (err) => {
      if (err) {
        console.error('保存文件时出错:', err);
      } else {
        console.log('文件保存成功!');
      }
    });
  } else {
    console.log('用户取消了保存操作');
  }
});

```
This behavior only occurs in Electron applications, including those built using Electron, such as Visual Studio Code, QQ,WeChat PC (version 4.0) and other apps.
However, Windows built-in applications like Notepad or third-party editors such as Notepad++ do not exhibit this behavior. In these apps, once illegal characters are removed from the file name, the modified name is accepted without requiring a file extension.

This inconsistency suggests that the showSaveDialog function in Electron has stricter handling or some kind of caching issue when validating file names, leading to unexpected behavior. This issue affects the user experience, as users are required to manually add file extensions even when they expect the file to save without one.

This issue seems to be related to or may have similarities with the following issue reported previously:
[#25615 - SaveDialog doesn’t allow saving a file with some extensions on Windows](https://github.com/electron/electron/issues/25615).

Both issues may indicate an underlying problem in how the dialog.showSaveDialog function interacts with Windows file system rules.

Image
Image
Image
Image

### PR DESCRIPTION
**Pull request by nilaoda**
#### Description of Change

Fix #44277

#### Release Notes

Notes: Fix update file extension bug

### CODE REVIEW

**Comment by ckerr** [comment]
> Applying: printing.patch
Patch failed at 0025 printing.patch
When you have resolved this problem, run "git am --continue".

This [win-x64-testing CI failure](https://ci.appveyor.com/project/electron-bot/electron-x64-testing/builds/50826296) doesn't seem to have anything do to with this PR. :thinking: I'm not sure how we got here unless `main` was somehow broken a couple of days ago when this PR was submitted...?

@nilaoda could you please rebase this PR onto electron/electron `main` to see if that resolves it?

**Comment by nilaoda** [comment]
@ckerr I pulled the main branch directly without making any modifications, and I encountered the same issue when compiling with @electron/build-tools. I don’t believe this is related to my MR.🥲

Additionally, could you use Squash merge when merging this PR? This would help keep the commit list clean.

File: shell/browser/ui/file_dialog_win.cc [hunk 1]
```diff
@@ -144,10 +144,14 @@ static void ApplySettings(IFileDialog* dialog, const DialogSettings& settings) {
...
-    if (std::wstring(filterspec[i].pszSpec) != L"*.*") {
+    std::wstring spec(filterspec[i].pszSpec);
+    // only need first extension
+    std::wstring firstSpec = spec.substr(0, spec.find(L';'));
+    if (firstSpec != L"*.*") {
       // SetFileTypeIndex is regarded as one-based index.
       dialog->SetFileTypeIndex(i + 1);
-      dialog->SetDefaultExtension(filterspec[i].pszSpec);
+      // "*.jpg" => "jpg"
+      dialog->SetDefaultExtension(firstSpec.substr(2).c_str());
```
Comment (samuelmaddock) [inline shell/browser/ui/file_dialog_win.cc:154]:
```suggestion
      // "*.jpg;*.png" => "*.jpg"
      std::wstring first_spec = spec.substr(0, spec.find(L';'));
      // "*.jpg" => "jpg"
      dialog->SetDefaultExtension(first_spec.substr(2).c_str());
```

File: shell/browser/ui/file_dialog_win.cc [hunk 2]
```diff
@@ -144,10 +144,14 @@ static void ApplySettings(IFileDialog* dialog, const DialogSettings& settings) {
   // We set file extension to the first none-wildcard extension to make
   // sure the dialog will update file extension automatically.
   for (size_t i = 0; i < filterspec.size(); ++i) {
-    if (std::wstring(filterspec[i].pszSpec) != L"*.*") {
+    std::wstring spec(filterspec[i].pszSpec);
+    // only need first extension
+    std::wstring firstSpec = spec.substr(0, spec.find(L';'));
+    if (firstSpec != L"*.*") {
```
Comment (samuelmaddock) [inline shell/browser/ui/file_dialog_win.cc:150]: In cases where `spec` is `*.*`, I think the substr will not be necessary.
```suggestion
    std::wstring spec(filterspec[i].pszSpec);
    if (spec != L"*.*") {
```

**Review by samuelmaddock — CHANGES_REQUESTED** [review]

File: shell/browser/ui/file_dialog_win.cc [hunk 3]
```diff
@@ -144,10 +144,14 @@ static void ApplySettings(IFileDialog* dialog, const DialogSettings& settings) {
...
   // sure the dialog will update file extension automatically.
   for (size_t i = 0; i < filterspec.size(); ++i) {
-    if (std::wstring(filterspec[i].pszSpec) != L"*.*") {
+    std::wstring spec(filterspec[i].pszSpec);
+    if (spec != L"*.*") {
       // SetFileTypeIndex is regarded as one-based index.
       dialog->SetFileTypeIndex(i + 1);
-      dialog->SetDefaultExtension(filterspec[i].pszSpec);
+      // "*.jpg;*.png" => "*.jpg"
+      std::wstring first_spec = spec.substr(0, spec.find(L';'));
```
Comment (samuelmaddock) [inline shell/browser/ui/file_dialog_win.cc:152]: does this still work when there's only one extension? as in "*.jpg" with no ';'

Comment (nilaoda) [reply shell/browser/ui/file_dialog_win.cc:152]: > does this still work when there's only one extension? as in "*.jpg" with no ';'

Yes, this code will still work when there’s only one extension (e.g., `"*.jpg"` without a `';'`). The `spec.find(L';')` call will return `std::wstring::npos` in that case, so `substr(0, std::wstring::npos)` will simply return the entire string, `"*.jpg"`. Thus, it handles both cases correctly.

**Review by samuelmaddock — APPROVED** [review]

**Comment by samuelmaddock** [comment]
@nilaoda can you rebase your PR? that way it'll include a fix for some of the failing tests in CI.

**Review by jkleinsc — APPROVED** [review]

**Pull request merged**